apiserver/HACluster.md

```diff
@@ -0,0 +1,215 @@
+# Creating HA cluster with API Server
+
+One of the issue for long-running Ray applications, for example, Ray Serve is that Ray Head node is a single
+point of failure, which means that if the Head node dies, complete cluster has to be restarted. Fortunately,
+KubeRay cluster provides an option to create
+[fault tolerance Ray cluster](https://docs.ray.io/en/master/cluster/kubernetes/user-guides/kuberay-gcs-ft.html).
+The similar type of highly available Ray cluster can also be created using API server. The foundation of this
+approach is ensuring high availability Global Control Service (GCS) data. GCS manages cluster-level
+metadata. By default, the GCS lacks fault tolerance as it stores all data in-memory, and a failure can cause the
+entire Ray cluster to fail. To make the GCS fault tolerant, you must have a high-availability Redis. This way,
+in the event of a GCS restart, it retrieves all the data from the Redis instance and resumes its regular
+functioning.
+
+## Creating external Redis cluster
+
+A comprehensive documentation on creating Redis cluster on Kubernetes can be found
+[here]( https://www.dragonflydb.io/guides/redis-kubernetes). For this example we will use a rather simple
+[yaml file](test/cluster/redis/redis.yaml). To create Redis run:
+
+```sh
+kubectl create ns redis
+kubectl apply -f <your location>/kuberay/apiserver/test/cluster/redis/redis.yaml -n redis
+```
+
+Note that here we are deploying redis to the `redis` namespace, that we are creating here.
+
+Alternatively, if you run on the cloud you can use managed version of HA Redis, which will not require
+you to stand up, run, manage and monitor your own version of redis.
+
+## Creating Redis password secret
+
+Before creating your cluster, you also need to create [secret](test/cluster/redis/redis_passwrd.yaml) in the
+namespace where you want to create your Ray cluster (remember, that secret is visible only within a given
+namespace). To create a secret for using external redis run:
+
+```sh
+kubectl apply -f <your location>/kuberay/apiserver/test/cluster/redis/redis_passwrd.yaml
+```
+
+## Ray Code for testing
+
+For both Ray jobs and Ray Serve we are recommending packaging the code in the image. For a simple testing here
+we will create a [config map](test/cluster/code_configmap.yaml), containg simple code, that we will use for
+testing. To deploy it run the following:
+
+```sh
+kubectl apply -f <your location>/kuberay/apiserver/test/cluster/code_configmap.yaml
+```
+
+## API server request
+
+To create a a cluster we can use the following curl command:
+
+```sh
+curl -X POST 'localhost:8888/apis/v1alpha2/namespaces/default/clusters' \
+--header 'Content-Type: application/json' \
```

Comment:
you might want to mention how the api server was started, in this case it was `make run`.

Reply:
Isn't this obvious?

Reply:
No, it is not. But we can fix this later.